src/actions/pipelines.js

```diff
@@ -0,0 +1,129 @@
+import { getUrl } from '../config';
+import loadJsonData from '../store/load-data';
+import { preparePipelineState } from '../store/initial-state';
+import { resetData } from './index';
+
+/**
+ * This file contains actions that update the active pipeline, and if loading data
+ * asynchronously, they also handle loading this pipeline data from different endpoints.
+ *
+ * Many different cases need to be addressed, including:
+ * 1. Loading data synchronously, or asynchonously.
+ * 2. Loading data and updating the pipeline on first page load, or on user actions.
+ * 3. Whether the dataset has pipelines defined in it, or not.
+ * 4. Whether localStorage has an active pipeline already defined.
+ * 5. If so, whether it exists in the current dataset.
+ * 6. Whether the requested endpoint is the 'main' one, or not.
+ * 7. Whether the pipeline flag is disabled.
+ *
+ * These are mostly handled either within this file, in the preparePipelineState
+ * utility, or in the getNodeDisabledPipeline selector. Please see their tests
+ * for more info about implementation requirements.
+ */
+
+export const UPDATE_ACTIVE_PIPELINE = 'UPDATE_ACTIVE_PIPELINE';
+
+/**
+ * Update the actively-selected pipeline
+ * @param {string} pipeline Pipeline ID
+ */
+export function updateActivePipeline(pipeline) {
+  return {
+    type: UPDATE_ACTIVE_PIPELINE,
+    pipeline
+  };
+}
+
+export const TOGGLE_PIPELINE_LOADING = 'TOGGLE_PIPELINE_LOADING';
+
+/**
+ * Toggle whether to display the loading spinner
+ * @param {boolean} loading True if pipeline is still loading
+ */
+export function toggleLoading(loading) {
+  return {
+    type: TOGGLE_PIPELINE_LOADING,
+    loading
+  };
+}
+
+/**
+ * Determine where to load data from
+ * @param {object} pipeline Pipeline state
+ */
+export const getPipelineUrl = pipeline => {
+  if (pipeline.active === pipeline.default) {
+    return getUrl('main');
+  }
+  return getUrl('pipeline', pipeline.active);
+};
+
+/**
+ * Check whether to make another async data request
+ * @param {object} pipeline Pipeline state
+ */
+export const requiresSecondRequest = (flags, pipeline) => {
```

Comment:
Is there a more specific name possible here? Something like `shouldRequestXXX`. If there are multiple things in the second request, you could add multiple functions with this naming style to make it clear what's needed per-case, even if they all share this same base function.

Reply:
In this case, it's just about whether it needs to make another pipeline endpoint request, i.e. to an endpoint other than 'main'. I don't have a specific name for these other endpoints so I avoided referring to them specifically because it would be more confusing? like, the best I can come up with is `shouldRequestAnotherEndpoint` which is garbage. I think it's implied that it's referring to pipeline endpoints because of what file it's in and which function is calling it. I think if anything, the best option here would be to add more detail to the JSDoc comment?

Reply:
Yeah I see how it could be difficult to make this code more self describing, so a comment is fine to avoid too much refactoring. Though I do think this updated comment could be simplified / split up to make it clearer, it's not a huge deal.

Reply:
Cool, I've updated the comments again